src-cljs/validation.cljs

```diff
@@ -0,0 +1,41 @@
+(ns doctopus.validation
+  (:require [clojure.string :as s]
+            [doctopus.util :refer [in?]]))
+
+(enable-console-print!)
+
+(defn field-empty?
+  "Returns true if contents is empty, whitespace trimmed."
+  [contents]
+  (empty? (s/trim contents)))
+
+(defn field-duplicate-value?
+  "Returns true if value exists under field in each values."
+  [field values value]
+  (in? value (map #(get % field) values)))
+
+(defn field-invalid-characters?
+  ([contents] (field-invalid-characters? contents #"[\w-_]+"))
+  ([contents rex]
+   (and (not= contents "") (->> contents
+                                 (re-matches rex)
+                                 nil?))))
+
+(defn create-form-validator
```

Comment:
I'm having feels about the flatten / for / for/ if-let / if combo, but I cannot yet see how to untangle it. Don't let this be a blocker -- I'll post something if I can get my head straight.